### PR TITLE
Add some "compile" tests for deriving Associations

### PR DESCRIPTION
**Pull request by weiznich**
Any future cases that should be added?

### CODE REVIEW

**Comment by sgrif** [comment]
I'd like us to test that these are actually generating the code we expect. We should execute the queries and expect to see some data back.

**Comment by killercup** [comment]
Thanks!

I fear this will fail CI with `must_use` errors, though. Let's add insert some data and assert the results.

**Comment by weiznich** [comment]
> I'd like us to test that these are actually generating the code we expect. We should execute the queries and expect to see some data back.

There are already tests doing this in diesel_test. So I'm not sure if we should add those tests on two location? 

> I fear this will fail CI with `must_use errors`, though. Let's add insert some data and assert the results.

There is an `#![allow(unused_must_use)]` at the top of the file to prevent this errors.

**Comment by killercup** [comment]
> There is an #![allow(unused_must_use)] at the top of the file to prevent this errors.

Lol, I apparently can't read. …And I've just been drinking tea today before anyone asks! 😅

**Comment by sgrif** [comment]
The only test I can see which executes a query involving a `foreign_key` option is the test for self referencing associations.

**Comment by sgrif** [comment]
We need to provide *some* tests for this behavior, even if it's not testing the actual results of executing the query. I'd like to at least test that `foreign_key` is doing the right thing, and probably a test like:

```rust
assert_eq!(
    debug_sql(Foo::belonging_to(&bar)),
    debug_sql(foos.filter(bar_id.eq(bar.id))),
);
```

**Comment by weiznich** [comment]
>  I'd like to at least test that foreign_key is doing the right thing, and probably a test like:

I've adjusted the diesel_codegen tests to check the generated sql

> The only test I can see which executes a query involving a foreign_key option is the test for self referencing associations.

I've added one test to diesel_tests for this scenario.

File: diesel_codegen/tests/associations.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        user_id: i32,
+        title: String,
+    }
+
+
+    posts::table.inner_join(users::table)
+        .select((users::id, users::name))
+        .filter(posts::id.eq(1)
+            .and(posts::user_id.eq(2))
+                .and(posts::title.eq("Bar")));
```
Comment (killercup) [inline diesel_codegen/tests/associations.rs]: Can you name these "compile-success assertions" by making them let statements? E.g., `let _can_select_columns_from_join = posts::table.inn…` (or a better name)

File: diesel_codegen/tests/associations.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+
+    #[derive(Associations, Identifiable)]
+    #[belongs_to(Tree, foreign_key = "parent_id")]
+    pub struct Tree {
+        id: i32,
+        parent_id: Option<i32>,
+    }
+    let t = Tree{ id: 42, parent_id: None };
```
Comment (killercup) [inline diesel_codegen/tests/associations.rs]: pls add space before `{` (sorry my brain zooms in on inconsistencies like that)

File: diesel_tests/tests/associations.rs [hunk 3]
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
...
+        id: i32,
+        name: String
+    }
+    #[derive(Debug, Queryable, Associations, Identifiable, Clone, PartialEq)]
+    #[belongs_to(User, foreign_key="belongs_to_user")]
+    pub struct Post {
+        id: i32,
+        belongs_to_user: i32,
+        title: String,
+    }
```
Comment (killercup) [inline diesel_tests/tests/associations.rs]: Please add empty lines before each of the above items where missing

File: diesel_tests/tests/associations.rs [hunk 4]
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
...
+
+    table! {
+        posts {
+            id -> Integer,
+            belongs_to_user -> Integer,
+            title -> Text,
+        }
+    }
+
+    #[derive(Debug, Queryable, Identifiable, PartialEq, Clone)]
```
Comment (killercup) [inline diesel_tests/tests/associations.rs]: Can you put diesel's derives last? Or… sort them alphabetically? Or anything that I can identify as 'order'? Or reveal the system you ordered them by? :D Sorry for commenting on all the nits

File: diesel_tests/tests/associations.rs [hunk 5]
```diff
@@ -187,3 +187,70 @@ fn conn_with_test_data() -> (TestConnection, User, User, User) {
...
+        belongs_to_user: i32,
+        title: String,
+    }
+    let sean = User { id: 1, name: "Sean".into() };
+    let tess = User { id: 2, name: "Tess".into() };
+    let post1 = Post { id: 1, belongs_to_user: 1, title: "Hello".into() };
+    let post2 = Post { id: 2, belongs_to_user: 2, title: "World".into() };
+    let post3 = Post { id: 3, belongs_to_user: 1, title: "Hello 2".into() };
+
+
```
Comment (killercup) [inline diesel_tests/tests/associations.rs]: ✂️

File: diesel_codegen/tests/associations.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+    let t = User { id: 42, name: "Sean".into() };
+
+    let belong_to = Post::belonging_to(&t);
+    let filter = posts::table.filter(posts::user_id.eq(42));
+
+    assert_eq!(
+        debug_sql::<Backend, _>(&belong_to),
+        debug_sql::<Backend, _>(&filter)
+    )
```
Comment (killercup) [inline diesel_codegen/tests/associations.rs]: The macro may return `()`, but in other tests we always append a `;`

**Comment by weiznich** [comment]
Any idea why the tests are failing? It builds fine for me locally using the same rust version (1.19.0)

**Pull request merged**